hal/src/common/gpio/v2/pin.rs

```diff
@@ -703,18 +711,186 @@ macro_rules! pin_id {
+/// Consequently, we repeat all of [`Pin`]'s inherenet methods on the trait, for
+/// convenience. To cast a generic type to a concrete type, use the
+/// [`as_concrete`](AnyPin::as_concrete) method. To cast back to the generic
+/// type, use the [`as_any`](Pin::as_any) method on [`Pin`].
 ///
-/// All instances of [Pin] implement this trait. When used as a trait bound, it
-/// acts to encapsulate a [Pin]. Without this trait, a completely generic
-/// [Pin] would require two type parameters. When using this trait as a bound,
-/// only one type parameter is required, yet you can still recover each type
-/// parameter of the corresponding [Pin] through the associated types.
-pub trait AnyPin: Sealed {
-    /// [PinId] of the corresponding [Pin]
+/// The following example walks through a few different ways to interact with a
+/// generic type constrained by `AnyPin`.
+///
+/// Suppose you wanted to store a completely generic [`Pin`] within a strcut.
+/// You can do so using only one type parameter and the [`AnyPin`] trait.
+///
+/// ```rust
+/// pub struct Example<P: AnyPin> {
+///     pin: P,
+/// }
+/// ```
+///
+/// Next, suppose you want to create a method that will take the [`Pin`] out of
+/// the struct, perform some operations in several [`PinMode`]s, and put it back
+/// into the struct before returning. The `elided` method below shows such an
+/// example. However, it can be a bit tricky to follow all of the type
+/// conversions here. For clarity, the `expanded` method shows the same behavior
+/// with each transformation given its proper type annotation. Notice that it is
+/// not enough to simply put back a type that implements
+/// `AnyPin<Id = P::Id, Mode = P::Mode>`. As far as the compiler is concerned,
+/// there could be many different types that satisfy that trait bound. Instead,
+/// you must put back a `P` exactly. The [`as_any`](Pin::as_any) function is the
+/// key here. It transforms some type that implements
+/// `AnyPin<Id = P::Id, Mode = P::Mode>` into `P` itself.
+///
+/// ```rust
+/// impl<P: AnyPin> Example<P> {
+///     pub fn elided(mut self) -> Self {
+///         let mut pin = self.pin.into_push_pull_output();
+///         pin.set_high().unwrap();
+///         let pin = pin.into_floating_input();
+///         let _bit = pin.is_low().unwrap();
+///         self.pin = pin.into_mode::<P::Mode>().as_any();
+///         self
+///     }
+///     pub fn expanded(mut self) -> Self {
+///         // ConcretePin<P> == Pin<P::Id, P::Mode>
+///         let pin: ConcretePin<P> = self.pin.as_concrete();
+///         let mut pin: Pin<P::Id, PushPullOutput> = pin.into_push_pull_output();
+///         pin.set_high().unwrap();
+///         let pin: Pin<P::Id, FloatingInput> = pin.into_floating_input();
+///         let _bit = pin.is_low().unwrap();
+///         let pin: Pin<P::Id, P::Mode> = pin.into_mode::<P::Mode>();
+///         self.pin: P = pin.as_any::<P>();
+///         self
+///     }
+/// }
+/// ```
```

Comment:
@ianrrees or someone else. Could you please have a look at this documentation and make sure it all makes sense?

Reply:
Looks good!  Just a few typos: inherenet, strcut, provies.